media-tv/kodi-9999: Use cmake instead of autotools

candrews opened the pull request:
Upstream is dropping support for autotools and requiring the use of cmake:
see xbmc/PR10429

https://bugs.gentoo.org/show_bug.cgi?id=601738

candrews commented:
I'm not a cmake or autotools expert by any means. However, I think I've set this up correctly (using Gentoo documentation and other ebuilds as guidance) but it doesn't work. Here's the error I'm getting:

```
>>> Working in BUILD_DIR: "/var/tmp/portage/media-tv/kodi-9999/work/kodi-9999_build"
cmake -C /var/tmp/portage/media-tv/kodi-9999/work/kodi-9999_build/gentoo_common_config.cmake -G Unix Makefiles -DCMAKE_INSTALL_PREFIX=/usr -DENABLE_CACHE=OFF -DENABLE_INTERNAL_FFMPEG=OFF -DENABLE_INTERNAL_CROSSGUID=OFF -DENABLE_NONFREE=no -DENABLE_ALSA=yes -DENABLE_AIRTUNES=yes -DENABLE_BLURAY=yes -DHAVE_LIBCAP=yes -DENABLE_CEC=no -DENABLE_DVDCSS=yes -DENABLE_OPTICAL=no -DENABLE_DBUS=yes -DENABLE_OPENGLES=no -DENABLE_LIRC=yes -DENABLE_MYSQLCLIENT=no -DENABLE_NFS=no -DENABLE_OPENGL=yes -DENABLE_OPENSSL=yes -DENABLE_PULSEAUDIO=yes -DENABLE_SMBCLIENT=yes -DENABLE_SDL=yes -DENABLE_SSH=yes -DENABLE_LIBUSB=yes -DENABLE_UPNP=yes -DENABLE_VAAPI=yes -DENABLE_VDPAU=no -DENABLE_MICROHTTPD=yes -DENABLE_X11=yes -DENABLE_XSLT=no -DENABLE_AVAHI=yes -DCMAKE_BUILD_TYPE=Release -DCMAKE_USER_MAKE_RULES_OVERRIDE=/var/tmp/portage/media-tv/kodi-9999/work/kodi-9999_build/gentoo_rules.cmake -DCMAKE_TOOLCHAIN_FILE=/var/tmp/portage/media-tv/kodi-9999/work/kodi-9999_build/gentoo_toolchain.cmake  /var/tmp/portage/media-tv/kodi-9999/work/kodi-9999/project/cmake/
loading initial cache file /var/tmp/portage/media-tv/kodi-9999/work/kodi-9999_build/gentoo_common_config.cmake
-- The CXX compiler identification is GNU 6.2.0
-- The C compiler identification is GNU 6.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/x86_64-pc-linux-gnu-gcc
-- Check for working CXX compiler: /usr/bin/x86_64-pc-linux-gnu-g++
-- Check for working CXX compiler: /usr/bin/x86_64-pc-linux-gnu-g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /usr/bin/x86_64-pc-linux-gnu-gcc
-- Check for working C compiler: /usr/bin/x86_64-pc-linux-gnu-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Source directory: /var/tmp/portage/media-tv/kodi-9999/work/kodi-9999
-- Build directory: /var/tmp/portage/media-tv/kodi-9999/work/kodi-9999_build
-- Generator: Single-configuration: Release (Unix Makefiles)
-- CMake Version: 3.7.1
-- System type: Linux
CMake Error at scripts/common/ArchSetup.cmake:71 (message):
  Toolchain did not define DEPENDS_PATH or NATIVEPREFIX.  Possibly outdated
  depends.
Call Stack (most recent call first):
  CMakeLists.txt:28 (include)


-- Configuring incomplete, errors occurred!
See also "/var/tmp/portage/media-tv/kodi-9999/work/kodi-9999_build/CMakeFiles/CMakeOutput.log".
 * ERROR: media-tv/kodi-9999::x-portage failed (configure phase):
 *   cmake failed
 * 
 * Call stack:
 *     ebuild.sh, line  115:  Called src_configure
 *   environment, line 5359:  Called cmake-utils_src_configure
 *   environment, line 1616:  Called _cmake_execute_optionally 'src_configure'
 *   environment, line  558:  Called enable_cmake-utils_src_configure
 *   environment, line 2062:  Called die
 * The specific snippet of code:
 *       "${CMAKE_BINARY}" "${cmakeargs[@]}" "${CMAKE_USE_DIR}" || die "cmake failed";
```

Could someone who knows a bit more about cmake and its use in ebuilds help me out?

candrews commented:
@mgorny based on the git log, you seem to be the authority on cmake-utils.eclass, and that's what's causing me confusion right now.

The cmake command run by the ebuild is:
```
cmake -C /var/tmp/portage/media-tv/kodi-9999/work/kodi-9999_build/gentoo_common_config.cmake -G "Unix Makefiles" -DCMAKE_INSTALL_PREFIX=/usr -DENABLE_CACHE=OFF -DENABLE_INTERNAL_FFMPEG=OFF -DENABLE_INTERNAL_CROSSGUID=OFF -DENABLE_NONFREE=no -DENABLE_ALSA=yes -DENABLE_AIRTUNES=yes -DENABLE_BLURAY=yes -DHAVE_LIBCAP=yes -DENABLE_CEC=no -DENABLE_DVDCSS=yes -DENABLE_OPTICAL=no -DENABLE_DBUS=yes -DENABLE_OPENGLES=no -DENABLE_LIRC=yes -DENABLE_MYSQLCLIENT=no -DENABLE_NFS=no -DENABLE_OPENGL=yes -DENABLE_OPENSSL=yes -DENABLE_PULSEAUDIO=yes -DENABLE_SMBCLIENT=yes -DENABLE_SDL=yes -DENABLE_SSH=yes -DENABLE_LIBUSB=yes -DENABLE_UPNP=yes -DENABLE_VAAPI=yes -DENABLE_VDPAU=no -DENABLE_MICROHTTPD=yes -DENABLE_X11=yes -DENABLE_XSLT=no -DENABLE_AVAHI=yes -DCMAKE_BUILD_TYPE=Release -DCMAKE_USER_MAKE_RULES_OVERRIDE=/var/tmp/portage/media-tv/kodi-9999/work/kodi-9999_build/gentoo_rules.cmake -DCMAKE_TOOLCHAIN_FILE=/var/tmp/portage/media-tv/kodi-9999/work/kodi-9999_build/gentoo_toolchain.cmake  /var/tmp/portage/media-tv/kodi-9999/work/kodi-9999/project/cmake/
```
which gives the error above.

However, by removing the additions made by cmake-utils I get this command:
```
cmake -G "Unix Makefiles" -DCMAKE_INSTALL_PREFIX=/usr -DENABLE_CACHE=OFF -DENABLE_INTERNAL_FFMPEG=OFF -DENABLE_INTERNAL_CROSSGUID=OFF -DENABLE_NONFREE=no -DENABLE_ALSA=yes -DENABLE_AIRTUNES=yes -DENABLE_BLURAY=yes -DHAVE_LIBCAP=yes -DENABLE_CEC=no -DENABLE_DVDCSS=yes -DENABLE_OPTICAL=no -DENABLE_DBUS=yes -DENABLE_OPENGLES=no -DENABLE_LIRC=yes -DENABLE_MYSQLCLIENT=no -DENABLE_NFS=no -DENABLE_OPENGL=yes -DENABLE_OPENSSL=yes -DENABLE_PULSEAUDIO=yes -DENABLE_SMBCLIENT=yes -DENABLE_SDL=yes -DENABLE_SSH=yes -DENABLE_LIBUSB=yes -DENABLE_UPNP=yes -DENABLE_VAAPI=yes -DENABLE_VDPAU=no -DENABLE_MICROHTTPD=yes -DENABLE_X11=yes -DENABLE_XSLT=no -DENABLE_AVAHI=yes -DCMAKE_BUILD_TYPE=Release  /var/tmp/portage/media-tv/kodi-9999/work/kodi-9999/project/cmake/
```
which works.

Do you have any tips as to what next steps I can take to figure out why the gentoo additions cause this problem?

akva2 commented:
"sorry" for jumping in, but seeing i'm the autotools removal dude (in work clothes): this is not your fault. this is the kodi cmake buildsystem assuming that a toolchain file means you must be crosscompiling, and then things goes haywire for you obviously. this must be fixed before autotools removal.

notspiff commented:
please try with pr ref'd above.

hauihau commented:
I have some remarks:

-> combine use flags cdio and optical: ENABLE_OPTICAL requires CDIO (I've chosen dvd in my local overlay)

-> udev is preferred over usb (see project/cmake/CMakeLists.txt), so better force udev instead of usb

-> keep use flag texturepacker, or at least add its built time dependencies on lzo, libpng, giflib and virtual/jpeg to DEPEND

-> change udisks? ( sys-fs/udisks:0 ) to udisks? ( sys-fs/udisks )

-> remove "?? ( gles vaapi )" from REQUIRED_USE

-> remove toolchain-funcs from inherit 

-> remove all ac_cv_* from ebuild

-> remove export HELP2MAN from src_configure

-> modify src_unpack see http://bugs.gentoo.org/598150

-> add dev-db/sqlite to COMMON_DEPEND

-> raise ffmpeg dep to >=3.1.6

-> move app-arch/bzip2, app-arch/unzip and app-arch/zip from COMMOND_DEPEND to DEPEND (need to check if these are still needed)

-> remove "\" from mycmakeargs (this should be an array)

-> add cmake option for bluetooth to mycmakeargs

-> remove many more unneeded packages:
 - app-i18n/enca
 - dev-python/simplejson
 - dev-python/pillow
 - media-fonts/corefonts
 - media-libs/flac
 - media-libs/jasper
 - media-libs/jbigkit
 - media-libs/libmad
 - media-libs/libmodplug
 - media-libs/libmpeg2
 - media-libs/libsamplerate
 - media-libs/tiff
 - media-sound/wavpack
 - media-libs/glew
 - x11-apps/xdpyinfo
 - x11-apps/mesa-progs
 - x11-libs/libXinerama
 - x11-proto/xineramaproto

candrews commented:
> I have some remarks:
> 
> -> combine use flags cdio and optical: ENABLE_OPTICAL requires CDIO (I've chosen dvd in my local overlay)

Done
> 
> -> udev is preferred over usb (see project/cmake/CMakeLists.txt), so better force udev instead of usb
Sorry, can you clarify what change you're suggesting?
> 
> -> keep use flag texturepacker, or at least add its built time dependencies on lzo, libpng, giflib and virtual/jpeg to DEPEND

Based on my reading of the cmake setup, texturepacker isn't optional, so we can't have a use flag controlling it. libpng is the only dependency you mentioned that wasn't already present, so I've added that (I also double checked - kodi does use it so it definitely should be there).
> 
> -> change udisks? ( sys-fs/udisks:0 ) to udisks? ( sys-fs/udisks )

Done
> 
> -> remove "?? ( gles vaapi )" from REQUIRED_USE

Done
> 
> -> remove toolchain-funcs from inherit

Done
> 
> -> remove all ac_cv_* from ebuild

Done
> 
> -> remove export HELP2MAN from src_configure

Done
> 
> -> modify src_unpack see BZ#598150

Done
> 
> -> add dev-db/sqlite to COMMON_DEPEND

I believe that's covered by `PYTHON_REQ_USE="sqlite"`, but I think what you're saying also makes sense. Done.
> 
> -> raise ffmpeg dep to >=3.1.6

Done
> 
> -> move app-arch/bzip2, app-arch/unzip and app-arch/zip from COMMOND_DEPEND to DEPEND (need to check if these are still needed)

Makes sense. From looking at the kodi source, I agree these aren't used at runtime. Done.
> 
> -> remove \ from mycmakeargs (this should be an array)

Done.
> 
> -> add cmake option for bluetooth to mycmakeargs

There is no cmake argument to enable/disable bluetooth (ENABLE_BLUETOOTH isn't an option). Do you have a suggestion as to how to accomplish this?

> 
> -> -DENABLE_CACHE=OFF should read -DENABLE_CCACHE=OFF

Done.
> 
> -> remove many more unneeded packages (BZ#598460):

Also dropped dependencies on media-libs/libmpeg2, media-libs/tiff, and media-libs/jasper.
Added dependency on pulseaudio? ( media-sound/pulseaudio )
> 
>     app-i18n/enca

Done
>     dev-python/simplejson

Done
>     dev-python/pillow

I believe pillow is used - can you please check again?
>     media-fonts/corefonts

Are you sure corefonts are unnecessary?
>     media-libs/flac

Done.
>     media-libs/jasper

Done.
>     media-libs/jbigkit

Done
>     media-libs/libmad

Done
>     media-libs/libmodplug
Done
>     media-libs/libmpeg2
Done (dropped in https://github.com/xbmc/xbmc/pull/8687 )
>     media-libs/libsamplerate
Done
>     media-libs/tiff
Done
>     media-sound/wavpack

Looks like this may be used... https://github.com/xbmc/xbmc/blob/17.0b6-Krypton/xbmc/music/tags/TagLoaderTagLib.cpp#L46 Thoughts?
>     media-libs/glew

Looks like this may be used... https://github.com/xbmc/xbmc/blob/17.0b6-Krypton/tools/depends/xbmc-addons.include#L93 Thoughts?
>     x11-apps/xdpyinfo

Done
>     x11-apps/mesa-progs

Done
>     x11-libs/libXinerama

Done
>     x11-proto/xineramaproto

Thank you very much for this feedback - I look forward to further feedback from you.

I've submitted some pull requests to Kodi requesting they update the documented dependencies:
https://github.com/xbmc/xbmc/pull/11111
https://github.com/xbmc/xbmc/pull/11112
https://github.com/xbmc/xbmc/pull/11113


Would you mind doing the same for any others as well as letting me know?

candrews commented:
@notspiff Thanks for jumping in - I'm using https://github.com/notspiff/xbmc/commit/578050baea3f618d83d914e4ce85e1feef6934ff as a patch, and made good progress. However, I'm now getting this error:
```
/usr/include/fribidi/fribidi-common.h:61:20: fatal error: glib.h: No such file or directory
```
Notably, glib.h exists and is available at /usr/include/glib-2.0/glib.h. Therefore, I believe there's a problem with https://github.com/xbmc/xbmc/blob/master/project/cmake/modules/FindFribidi.cmake and how it sets up the include dir - can you help with that issue?

hauihau commented:
>-> udev is preferred over usb (see project/cmake/CMakeLists.txt), so better force udev instead of usb
>Sorry, can you clarify what change you're suggesting?

The current ebuild hast +usb. But UDEv will be used, if it is found, because according to CMakeLists.txt, LibUSB is only searched, if UDEV_FOUND is false:
if(NOT UDEV_FOUND)
  core_optional_dep(LibUSB)
endif()

Note: configure.ac does the same,:
\# if libudev is available, we don't need libusb
if test "x$use_libudev" != "xno"; then
  use_libusb="no"
  AC_MSG_NOTICE($libusb_disabled_udev_found)
...


>-> keep use flag texturepacker, or at least add its built time dependencies on lzo, libpng, giflib and virtual/jpeg to DEPEND
>Based on my reading of the cmake setup, texturepacker isn't optional, so we can't have a use flag controlling it.

I guess that is something like crossguid, perhaps there may be a standalone texturepacker. But I agree, as long as this isn't the case, we should hard depend on the texturepacker deps but only fuer built time.


>libpng is the only dependency you mentioned that wasn't already present, so I've added that (I also double checked - kodi does use it so it definitely should be there).

I can't find any place where kodi uses libpng:
htpc kodi-9999 # grep -r png.h
tools/depends/native/TexturePacker/src/decoder/PNGDecoder.cpp:#include "png.h"


>-> add dev-db/sqlite to COMMON_DEPEND
>I believe that's covered by PYTHON_REQ_USE="sqlite", but I think what you're saying also makes sense. Done.

I have no clue why kodi depends on python[sqlite]. Kodi itself is using sqlite as the default database backend if I'm correct. It also has it as required_dep in CMakeLists.txt.


>-> add cmake option for bluetooth to mycmakeargs
>There is no cmake argument to enable/disable bluetooth (ENABLE_BLUETOOTH isn't an option). I think we just need a conditional dependency on dev-python/pybluez.

You're right. Only configure.ac has a switch for bluetooth. There is a FindBluetooth.cmake file, but it is not used. Perhaps we should report that to the kodi guys.

../xbmc/network/TCPServer.cpp has a #ifdef HAVE_LIBBLUETOOTH which is only set by configure.ac, while there is no occurence in project/cmake


> dev-python/pillow
>I believe pillow is used - can you please check again?

I've removed the dep some time ago and kodi builds just fine


> media-fonts/corefonts
>Are you sure corefonts are unnecessary?

I've removed the dep some time ago and kodi builds just fine


> media-sound/wavpack
> Looks like this may be used... https://github.com/xbmc/xbmc/blob/17.0b6-Krypton/xbmc/music/tags/TagLoaderTagLib.cpp#L46 Thoughts?

The include is provided by taglib:
htpc kodi-9999 # qfile /usr/include/taglib/wavpackfile.h
media-libs/taglib (/usr/include/taglib/wavpackfile.h)


> media-libs/glew
> Looks like this may be used... https://github.com/xbmc/xbmc/blob/17.0b6-Krypton/tools/depends/xbmc-addons.include#L93 Thoughts?

I don't know if this file is also used in cmake builds, I can find it only in a Makefile:
htpc kodi-9999 # grep -r xbmc-addons.include .
./tools/depends/target/binary-addons/Makefile:include ../../xbmc-addons.include


> I've submitted some pull requests to Kodi requesting they update the documented dependencies:
> xbmc/xbmc#11111
> xbmc/xbmc#11112
> xbmc/xbmc#11113

Thanks!


> Would you mind doing the same for any others as well as letting me know?

Which others do you mean?

Perhaps we should ask kodi developers about pillow, wavpack, corefonts and glew. Even though wavpack and glew ar einstalled on my system, they are not used:
htpc kodi-9999 # grep -ir wavpack /var/db/pkg/media-tv/kodi-9999/NEEDED
htpc kodi-9999 # grep -ir glew /var/db/pkg/media-tv/kodi-9999/NEEDED
htpc kodi-9999 # strings /usr/lib64/kodi/kodi.bin | egrep -i '(wavpack|glew)' | c++filt 
handle_708_TGW_ToggleWindows(cc708_service_decoder*, int)
TagLib::WavPack::File::File(TagLib::IOStream*, bool, TagLib::AudioProperties::ReadStyle)
TagLib::WavPack::File::APETag(bool)
togglewatched
ToggleWindows
htpc kodi-9999 #

hudokkow commented:
@candrews , fribidi problem solved by https://github.com/xbmc/xbmc/pull/11116

candrews commented:
> The current ebuild hast +usb. But UDEv will be used, if it is found, because according to CMakeLists.txt, LibUSB is only searched, if UDEV_FOUND is false:
> if(NOT UDEV_FOUND)
> core_optional_dep(LibUSB)
> endif()

I see what you're saying, thanks for the clarification. Perhaps the `usb` and `udev` flags should conflict then... that may be annoying for users though. Another option is to have the libusb only pulled in if `usb` is set and `udev` is not. What are your thoughts?

> I can't find any place where kodi uses libpng

I agree, I've removed the dependency on libpng.

> I've removed the dep some time ago and kodi builds just fine

Yes, but does it work right at run time? :-)



> media-sound/wavpack
>     Looks like this may be used... https://github.com/xbmc/xbmc/blob/17.0b6-Krypton/xbmc/music/tags/TagLoaderTagLib.cpp#L46 Thoughts?
> 
> The include is provided by taglib:
> htpc kodi-9999 # qfile /usr/include/taglib/wavpackfile.h
> media-libs/taglib (/usr/include/taglib/wavpackfile.h)

Removed that dependency.

> media-libs/glew
>     Looks like this may be used... https://github.com/xbmc/xbmc/blob/17.0b6-Krypton/tools/depends/xbmc-addons.include#L93 Thoughts?
> 
> I don't know if this file is also used in cmake builds, I can find it only in a Makefile:
> htpc kodi-9999 # grep -r xbmc-addons.include .
> ./tools/depends/target/binary-addons/Makefile:include ../../xbmc-addons.include

glew was removed from kodi: https://github.com/xbmc/xbmc/pull/8631
I've submitted a PR to remove the remaining references to it:
https://github.com/xbmc/xbmc/pull/11128

cjmayo commented:
Thanks to the work here I've now been able to build kodi-17.0_beta6 with cmake, although I had to add USE="webserver" (more on that at the end).
Also looks like most of the skin pngs are being built but not installed e.g. DefaultVideo.png
A few notes on the build and the discussion here.


+ For corefonts, arial.ttf is found in the source (perhaps it isn't critical to have it?):
```
$ grep -rn arial.ttf
addons/skin.estuary/xml/Font.xml:134:			<filename>arial.ttf</filename>
addons/skin.estuary/xml/Font.xml:139:			<filename>arial.ttf</filename>
...
system/settings/settings.xml:669:          <default>arial.ttf</default>
xbmc/cores/VideoPlayer/VideoRenderers/DebugRenderer.cpp:124:      text->PrepareRender("arial.ttf", 1, 16, 0, m_font, m_fontBorder);
xbmc/guilib/GUIFontManager.cpp:140:      if (strFilename != "arial.ttf")
xbmc/guilib/GUIFontManager.cpp:142:        CLog::Log(LOGERROR, "Couldn't load font name: %s(%s), trying to substitute arial.ttf", strFontName.c_str(), strFilename.c_str());
xbmc/guilib/GUIFontManager.cpp:143:        return LoadTTF(strFontName, "arial.ttf", textColor, shadowColor, iSize, iStyle, border, lineSpacing, originalAspect);
xbmc/utils/Splash.cpp:75:      auto messageFont = g_fontManager.LoadTTF("__splash__", "arial.ttf", 0xFFFFFFFF, 0, 20, FONT_STYLE_NORMAL, false, 1.0f, 1.0f, &res);
```

+ udisks is used over dbus, and it is the original "org.freedesktop.UDisks" not "org.freedesktop.UDisks2":
https://github.com/xbmc/xbmc/blob/master/xbmc/storage/linux/UDisksProvider.cpp#L281
https://bugs.gentoo.org/show_bug.cgi?id=503862

+ Haven't tested this on master, but
https://github.com/xbmc/xbmc/pull/11102
at least my kodi-17.0_beta6 is using libpng:
```
$ ldd /usr/lib64/kodi/kodi.bin | grep png
	libpng16.so.16 => /usr/lib64/libpng16.so.16
```
Messages during build:
```
-- Found PNG: /usr/lib64/libpng.so (found version "1.6.26+apng") 

/var/tmp/portage/media-tv/kodi-17.0_beta6-r100/work/kodi-17.0_beta6_build/addons/skin.estuary/media/Textures.xbt -dupecheck
This is a PNG - lets load it via libpng...
DefaultActor.png                              
yes
```

+ Use existing PATCHES instead of epatch?

+ With kodi-17.0_beta6 to stop library download from GitHub add to mycmakeargs:
```
		-Dlibdvdread_URL="${DISTDIR}/libdvdread-${LIBDVDREAD_COMMIT}.tar.gz" \
		-Dlibdvdnav_URL="${DISTDIR}/libdvdnav-${LIBDVDNAV_COMMIT}.tar.gz" \
		-Dlibdvdcss_URL="${DISTDIR}/libdvdcss-${LIBDVDCSS_COMMIT}.tar.gz" \
```

Then src_unpack() can be removed.

+ src_install() needs cmake-utils_src_install rather than default

+ Needed to patch out use of gtk-update-icon-cache:
https://github.com/cjmayo/xbmc/commit/5c3762dce112cf2c297c2c32fd494f00bc2c4191
(Is that a good idea for non-Gentoo users?)

+ Already had net-libs/libmicrohttpd installed with USE="-webserver":
```
-- MICROHTTPD enabled: No
....
build/network/httprequesthandler/python/network_httprequesthandlers_python.a(HTTPPythonWsgiInvoker.cpp.o):HTTPPythonWsgiInvoker.cpp:PythonModules: error: undefined reference to 'PythonBindings::initModule_xbmcwsgi()'
build/network/httprequesthandler/network_httprequesthandlers.a(HTTPRequestHandlerUtils.cpp.o):HTTPRequestHandlerUtils.cpp:function HTTPRequestHandlerUtils::GetRequestHeaderValue(MHD_Connection*, MHD_ValueKind, std::string const&): error: undefined reference to 'MHD_lookup_connection_value'
build/network/httprequesthandler/network_httprequesthandlers.a(HTTPRequestHandlerUtils.cpp.o):HTTPRequestHandlerUtils.cpp:function HTTPRequestHandlerUtils::GetRequestHeaderValues(MHD_Connection*, MHD_ValueKind, std::map<std::string, std::string, std::less<std::string>, std::allocator<std::pair<std::string const, std::string> > >&): error: undefined reference to 'MHD_get_connection_values'
build/network/httprequesthandler/network_httprequesthandlers.a(HTTPRequestHandlerUtils.cpp.o):HTTPRequestHandlerUtils.cpp:function HTTPRequestHandlerUtils::GetRequestHeaderValues(MHD_Connection*, MHD_ValueKind, std::multimap<std::string, std::string, std::less<std::string>, std::allocator<std::pair<std::string const, std::string> > >&): error: undefined reference to 'MHD_get_connection_values'
```

hauihau commented:
> I see what you're saying, thanks for the clarification. Perhaps the usb and udev flags should conflict then... that may be annoying for users though. Another option is to have the libusb only pulled in if usb is set and udev is not. What are your thoughts?

I have no idea how to deal with that. Perhaps there is some default behaviour which gentoo prefers.


>media-fonts/corefonts
>Are you sure corefonts are unnecessary?
>I've removed the dep some time ago and kodi builds just fine
>Yes, but does it work right at run time? :-)

Haven't seen any change in behaviour so far.

hauihau commented:
> Also looks like most of the skin pngs are being built but not installed e.g. DefaultVideo.png
A few notes on the build and the discussion here.

Not 100% sure, but this maybe because texturepacker packs all these images in one file.


> For corefonts, arial.ttf is found in the source (perhaps it isn't critical to have it?):
$ grep -rn arial.ttf
addons/skin.estuary/xml/Font.xml:134:			<filename>arial.ttf</filename>
addons/skin.estuary/xml/Font.xml:139:			<filename>arial.ttf</filename>
...

Thats just an additional fontset defined which uses arial instead of noto.

> system/settings/settings.xml:669:          <default>arial.ttf</default>

Looks like arial is used as default for subtitles, so the depencendy on corefonts should stay in my opinion. Arial is also used if the configured font can not be found/loaded (see  xbmc/guilib/GUIFontManager.cpp)

>udisks is used over dbus, and it is the original "org.freedesktop.UDisks" not "org.freedesktop.UDisks2":
https://github.com/xbmc/xbmc/blob/master/xbmc/storage/linux/UDisksProvider.cpp#L281
https://bugs.gentoo.org/show_bug.cgi?id=503862

Didn't see that, thanks. Perhaps also a candidate for upstream kodi to look at.

>Haven't tested this on master, but
xbmc/xbmc#11102
at least my kodi-17.0_beta6 is using libpng:

$ ldd /usr/lib64/kodi/kodi.bin | grep png
	libpng16.so.16 => /usr/lib64/libpng16.so.16

This comes as a dependency from another used shared lib:
htpc kodi-9999 # lddtree /usr/lib64/kodi/kodi.bin 
kodi.bin => /usr/lib64/kodi/kodi.bin (interpreter => /lib64/ld-linux-x86-64.so.2)
[snip]
    libpcre.so.1 => /lib64/libpcre.so.1
    libfreetype.so.6 => /usr/lib64/libfreetype.so.6
        libbz2.so.1 => /lib64/libbz2.so.1
        libpng16.so.16 => /usr/lib64/libpng16.so.16
    libtag.so.1 => /usr/lib64/libtag.so.1
[snip]

> Messages during build:

-- Found PNG: /usr/lib64/libpng.so (found version "1.6.26+apng") 

Yes, because tools/depends/native/TexturePacker/CMakeLists.txt requires it:
find_package(PNG REQUIRED)


> Already had net-libs/libmicrohttpd installed with USE="-webserver":

-- MICROHTTPD enabled: No
....
build/network/httprequesthandler/python/network_httprequesthandlers_python.a(HTTPPythonWsgiInvoker.cpp.o):HTTPPythonWsgiInvoker.cpp:PythonModules: error: undefined reference to 'PythonBindings::initModule_xbmcwsgi()'
build/network/httprequesthandler/network_httprequesthandlers.a(HTTPRequestHandlerUtils.cpp.o):HTTPRequestHandlerUtils.cpp:function HTTPRequestHandlerUtils::GetRequestHeaderValue(MHD_Connection*, MHD_ValueKind, std::string const&): error: undefined reference to 'MHD_lookup_connection_value'
build/network/httprequesthandler/network_httprequesthandlers.a(HTTPRequestHandlerUtils.cpp.o):HTTPRequestHandlerUtils.cpp:function HTTPRequestHandlerUtils::GetRequestHeaderValues(MHD_Connection*, MHD_ValueKind, std::map<std::string, std::string, std::less<std::string>, std::allocator<std::pair<std::string const, std::string> > >&): error: undefined reference to 'MHD_get_connection_values'
build/network/httprequesthandler/network_httprequesthandlers.a(HTTPRequestHandlerUtils.cpp.o):HTTPRequestHandlerUtils.cpp:function HTTPRequestHandlerUtils::GetRequestHeaderValues(MHD_Connection*, MHD_ValueKind, std::multimap<std::string, std::string, std::less<std::string>, std::allocator<std::pair<std::string const, std::string> > >&): error: undefined reference to 'MHD_get_connection_values'

Looks like an error which upstream kodi should fix.

hauihau commented:
@candrews:

- please move libpng, giflib and virtual/jpeg too DEPEND (they aren't needed by kodi itself, only TexturePacker uses them, so DEPEND is sufficient).
- please revert the change to pybluez, bluez should stay, see xbmc/xbmc/pull/11137
- please remove media-fonts/anonymous-pro and media-fonts/dejavu from COMMON_DEPEND

hudokkow commented:
We are giving you guys a run for your money. :wink:
Thanks for finding the bugs. We're here to squash'em. Hopefully...

hauihau commented:
@hudokkow:

could you please tell us about:
- migration plans to UDisks2
- is pillow needed
- is corefonts needed
- the undefined references when ENABLE_MICROHTTPD=off but libmicrohttpd is installed

candrews commented:
> how to pass/enable HAVE_LIBCAP

I think that's a bug in the Kodi, I reported it at http://trac.kodi.tv/ticket/17130

notspiff commented:
> I have no clue why kodi depends on python[sqlite]. Kodi itself is using sqlite as the default database 
> backend if I'm correct. It also has it as required_dep in CMakeLists.txt.

it's due to legacy. lots of add-ons (used to) implicitly assume it's available in the kodi-hosted python env without explicitly depending on the script module. since we run on the system interpreter, it needs to be around on the system level.

hauihau commented:
Sorry, it's me again:
- the exported HAVE_GIT is only used in configure.ac
- GIT_REV is used a line output header in ./project/cmake/CMakeLists.txt: message(STATUS "GIT_REV: ${APP_SCMID}")

(Dont' know what the guidlines to sorting in *DEPEND are)
- sort IUSE alphabetically after renaming usb to libusb
- sort REQUIRED_USE (udev comes before udisks)
- sort packages in COMMON_DEPEND/DEPEND

 alphabetically
- sort use flags in mycmakearg alphabetically

- use the updated patch from xbmc/xbmc/pull/11088
- -DENABLE_UDEV=$(usex udev) missing in mycmakearg
- -DENABLE_LIBUSB is still reported as Manually-specified variable, upstream kodi should take a look on that
- is the sed in src_prepare still needed? 

I've attached my current ebuild diff [kodi-9999.patch.txt](https://github.com/gentoo/gentoo/files/645959/kodi-9999.patch.txt)

wsnipex commented:
for reference: pillow is used by a few python addons. e.g. openweathermap

hauihau commented:
@wsnipex: thanks for your feedback. So it's not a dependency of kodi itself. If a plugin needs pillow/pil, it should mention it in its readme.

hauihau commented:
I managed to successfully compile kodi-9999 both with autotools (needed to patch configure.ac because of xbmc/xbmc/pull/10922) and cmake. MY findings so far:

**cmake** build lacks all compatitiblity symlinks:
- /usr/bin/xbmc
- /usr/bin/xbmc-standalone
- /usr/include/xbmc
- /usr/lib64/xbmc
- /usr/share/xbmc
- /usr/share/xsessions/xbmc.desktop

**cmake** build does not have .debug files (I'm using FEATURE=splitdebug) for
- all .so files in /usr/lib64/kodi/addons
- /usr/lib64/kodi/system/libsse4-x86_64-linux.so

**cmake** build has /usr/share/doc/kodi and /usr/share/doc/kodi-9999
```
htpc ~ # ls /var/tmp/kodi-*/image/usr/share/doc/*/
/var/tmp/kodi-autotools/image/usr/share/doc/kodi-9999/:
README.linux.bz2  README.md.bz2  version.txt.bz2

/var/tmp/kodi-cmake/image/usr/share/doc/kodi/:
README.linux.bz2  version.txt.bz2

/var/tmp/kodi-cmake/image/usr/share/doc/kodi-9999/:
README.md.bz2
```

**cmake** build does not install png files in media folder of skins:
```
htpc ~ # ls /var/tmp/kodi-*/image/usr/share/kodi/addons/skin.estouchy/media/
/var/tmp/kodi-autotools/image/usr/share/kodi/addons/skin.estouchy/media/:
arrow_down.png
arrow_left.png
[...]
icon_settings_services.png
icon_settings_systeminfo.png
icon_settings_system.png
kodi_logo.png
list_focus.png
Makefile
Makefile.in
osd_forward.png
osd_next.png
[...]
subtitles
Textures.xbt
thumb_focus.png
weather
white.png

/var/tmp/kodi-cmake/image/usr/share/kodi/addons/skin.estouchy/media/:
Textures.xbt

htpc ~ # ls /var/tmp/kodi-*/image/usr/share/kodi/addons/skin.estuary/media/
/var/tmp/kodi-autotools/image/usr/share/kodi/addons/skin.estuary/media/:
buttons
calibrate
colors
curial.xbt
DefaultActor.png
[...]
DefaultYear.png
dialogs
flags
flat.xbt
[...]
Makefile
Makefile.in
osd
[...]
Textures.xbt
weather
windows

/var/tmp/kodi-cmake/image/usr/share/kodi/addons/skin.estuary/media/:
curial.xbt
flat.xbt
Textures.xbt
```

**cmake** build installs .cmake files to /usr/share/kodi/cmake instead of /usr/lib64/kodi/
- autotools build has CMakeHelpers.cmake and Uninstall.cmake
- cmake build has PathSetup.cmake


**cmake** build exclusive files:
- /usr/bin/TexturePacker

**autotools**  build exclusive files:
- /usr/lib64/kodi/system/players/VideoPlayer/libdvdcss-x86_64-linux.so
- /usr/share/kodi/system/IRSSmap.xml
- /usr/share/kodi/system/python
- /usr/share/kodi/system/python/readme.txt
- /usr/share/kodi/system/X10-Lola-IRSSmap.xml

**cmake** build has a lot of duplicates in the final link command of kodi.bin

If 've attached the build.logs of both autotools [kodi-autotools-build.log.txt](https://github.com/gentoo/gentoo/files/647454/kodi-autotools-build.log.txt) and cmake [kodi-cmake-build.log.txt](https://github.com/gentoo/gentoo/files/647469/kodi-cmake-build.log.txt) builds.

cjmayo commented:
+ Double /usr/share/doc directories fixed with https://github.com/xbmc/xbmc/pull/11151 and:
`-Ddocdir=/usr/share/doc/${PF}`
Of course may not want all those files.

+ Replace -ne 9999 with != 9999 in:
```
	[[ ${PV} -ne 9999 ]] &&	mycmakeargs+=(
		-Dlibdvdread_URL="${DISTDIR}/libdvdread-${LIBDVDREAD_COMMIT}.tar.gz"
		-Dlibdvdnav_URL="${DISTDIR}/libdvdnav-${LIBDVDNAV_COMMIT}.tar.gz"
		-Dlibdvdcss_URL="${DISTDIR}/libdvdcss-${LIBDVDCSS_COMMIT}.tar.gz"
	)
```
else with PV=17.0_beta6:
`/var/tmp/portage/media-tv/kodi-17.0_beta6-r100/temp/environment: line 4769: [[: 17.0_beta6: syntax error: invalid arithmetic operator (error token is ".0_beta6")`

+ Replace double "/" in MY_PV with single:
```
	MY_PV=${MY_PV//_alpha/a}
	MY_PV=${MY_PV//_beta/b}
	MY_PV=${MY_PV//_rc/rc}
```

wsnipex commented on media-tv/kodi/kodi-9999.ebuild:
```diff
@@ -152,128 +150,79 @@ In some cases Kodi needs to access multicast addresses.
+		-DENABLE_OPTICAL=$(usex dvd)
+		-DENABLE_PLIST=$(usex airplay)
+		-DENABLE_PULSEAUDIO=$(usex pulseaudio)
+		-DENABLE_SDL=$(usex sdl)
```
SDL is not used on linux anymore

hauihau commented:
/xbmc/xbmc/pull/11088 has been merged

hauihau commented:
please also remove kodi-9999-no-arm-flags.patch from patches (it's meant for configure.ac)

hauihau commented:
hi craig,

- pillow and simplejson are back
- make gles? ( .. a oneliner
- lirc does not require app-misc/lirc, ENABLE_LIRC sets LIRC_DEVICE
- rename -DENABLE_LIBCAP to -DENABLE_CAP
- cosmetics: `[[ ${PV} != 9999 ]] &&  mycmakeargs+=(` is a TAB after `&&`
- I've changed IUSE xslt to +xslt

candrews commented:
> pillow and simplejson are back

I feel like they belong as dependencies.  As @wsnipex said earlier:
> for reference: pillow is used by a few python addons. e.g. openweathermap

Removing these dependencies risks breaking plugins users may be using that depend upon them. I'm not really comfortable doing that. Is there any way to prevent such breakage?

> make gles? ( .. a oneliner

Sure
> lirc does not require app-misc/lirc, ENABLE_LIRC sets LIRC_DEVICE

It doesn't 'require' lirc to build, but it makes sense to require lirc, otherwise the LIRC_DEVICE flag doesn't make a lot of sense. For that reason, I've moved lirc to RDEPEND.

> rename -DENABLE_LIBCAP to -DENABLE_CAP

Done

> cosmetics: [[ ${PV} != 9999 ]] && mycmakeargs+=( is a TAB after &&

Much nicer

> I've changed IUSE xslt to +xslt

I did that already too :)

hauihau commented:
>pillow and simplejson are back
I feel like they belong as dependencies. As @wsnipex said earlier:

> for reference: pillow is used by a few python addons. e.g. openweathermap
Removing these dependencies risks breaking plugins users may be using that depend upon them. I'm not really comfortable doing that. Is there any way to prevent such breakage?

I think that plugin devolpers should mention that in there plugins if they need pillow or simplejson. As they are no kodi dependencies, I would like to seethem gone. Perhpas an einfo would be tho correct thing.

> lirc does not require app-misc/lirc, ENABLE_LIRC sets LIRC_DEVICE
It doesn't 'require' lirc to build, but it makes sense to require lirc, otherwise the LIRC_DEVICE flag doesn't make a lot of sense. For that reason, I've moved lirc to RDEPEND.

It makes sense. I'm using inputlircd, which provides /dev/lircd. How about an einfo when use lirc is enabled, that one should install app-misc lirc or app-misc/inputlircd to make use of lirc?

wsnipex commented:
Python based kodi addons are installed directly from the kodi addon repository.
Addons specify their dependencies in their addon.xml and kodi makes sure those deps are available when the addon is installed. Pillow is something that we cannot install on demand, because it's compiled code, so it is shipped with kodi and assumed to be available on all platforms.

hauihau commented:
@wsnipex I got the point. By shipped you mean tools/depends/target/pythonmodule-pil?

But what about simplejson? And how does kodi handle addons which e.g require lxml, or beautifulsoup? Or any other of the many python modules out there? How would kodi ensure that these are available?

I've seen that there is also pythonmodule-setuptools in tools/depends/target/, so I guess this should also be handled like pillow and be a dep in the ebuild.

wsnipex commented:
everything that is pure python(doesn't need compiling) is provided in the kodi addon repository:
http://mirrors.kodi.tv/addons/krypton/ 
simplejson and beautifulsoup are among them.

Compiled ones like pillow have to be installed on the system, either shipped with kodi(windows, osx, android) or via package manager on linux, because on linux usually system python is used.

hauihau commented:
@wsnipex thanks for making this clear. So we can remove simplejson from the depends. But what about setuptools? Looks like it is also a requirement, correct?

@candrews so please rename simplejson to setputools.

wsnipex commented:
kodi doesn't require setuptools. We only use it to build stuff like pillow on platforms that do not use system python. So the ebuild should add a dependency on pillow, but not on setuptools.

candrews commented:
> It makes sense. I'm using inputlircd, which also provides /dev/lircd. How about an einfo when use lirc is enabled, that one needs to install app-misc lirc or app-misc/inputlircd to make use of lirc?

How about an || allowing either lirc or inputlircd to satisfy the dependency?

I'll keep pillow in and drop simplejson.

@wsnipex congratulations on dropping autotools! I see the PR was merged. So clearly this will be in Kodi 18, but will it be in Krypton (presumable 17.0b7)?

wsnipex commented:
thanks! Krypton will still have autotools.

hauihau commented:
>>It makes sense. I'm using inputlircd, which also provides /dev/lircd. How about an einfo when use lirc is enabled, that one needs to install app-misc lirc or app-misc/inputlircd to make use of lirc?

>How about an || allowing either lirc or inputlircd to satisfy the dependency?

I'm not sure how to deal with that. As kodi itself compiles fine without lirc or inputlircd installed, I would go with an einfo. If you look at other packages, they also use einfo if the user should install specific packages for additional functionality.

cjmayo commented:
Latest ideas...

+ dev-libs/libcec-4.0.0_p3 is now in the tree. libcec dependency does need upping to >= 4.0:
`-- Could NOT find CEC: Found unsuitable version "3.1.0", but required is at least "4.0.0" (found /usr/lib64/libcec.so)`

+ suggest adding `-DENABLE_AIRTUNES=OFF`, saves a failed check and avoids its warning:
`-- Could NOT find Shairplay (missing:  SHAIRPLAY_LIBRARY SHAIRPLAY_INCLUDE_DIR HAVE_SHAIRPLAY_CALLBACK_CLS)`

+ `-Ddocdir=/usr/share/doc/${PF}` is now supported

+ suggest stopping binaries being stripped:
```
QA Notice: Pre-stripped files found:
/usr/bin/TexturePacker
/usr/lib64/kodi/addons/library.xbmc.pvr/libXBMC_pvr-x86_64-linux.so
/usr/lib64/kodi/addons/library.xbmc.codec/libXBMC_codec-x86_64-linux.so
/usr/lib64/kodi/addons/library.xbmc.addon/libXBMC_addon-x86_64-linux.so
/usr/lib64/kodi/addons/library.kodi.peripheral/libKODI_peripheral-x86_64-linux.so
/usr/lib64/kodi/addons/library.kodi.inputstream/libKODI_inputstream-x86_64-linux.so
/usr/lib64/kodi/addons/library.kodi.guilib/libKODI_guilib-x86_64-linux.so
/usr/lib64/kodi/addons/library.kodi.audioengine/libKODI_audioengine-x86_64-linux.so
/usr/lib64/kodi/addons/library.kodi.adsp/libKODI_adsp-x86_64-linux.so
/usr/lib64/kodi/system/libsse4-x86_64-linux.so
/usr/lib64/kodi/kodi.bin
```
fix here:
https://github.com/cjmayo/xbmc/commit/244a4623c617afbe213a0818f5bd86faeb3986e7

I've created a gentoo branch there with the `gtk-update-icon-cache` commit as well, and attached them as one patch here:
[kodi-9999-cmake.patch.txt](https://github.com/gentoo/gentoo/files/658077/kodi-9999-cmake.patch.txt)


+ don't think `domenu` is needed in src_install(), already installed by cmake/scripts/linux/Install.cmake:
```
# Install desktop entry
install(FILES ${CORE_SOURCE_DIR}/tools/Linux/kodi.desktop
        DESTINATION ${datarootdir}/applications
        COMPONENT kodi)
```

candrews commented:
@cjmayo,

> suggest stopping binaries being stripped:
> 
> QA Notice: Pre-stripped files found:
> /usr/bin/TexturePacker
> /usr/lib64/kodi/addons/library.xbmc.pvr/libXBMC_pvr-x86_64-linux.so
> /usr/lib64/kodi/addons/library.xbmc.codec/libXBMC_codec-x86_64-linux.so
> /usr/lib64/kodi/addons/library.xbmc.addon/libXBMC_addon-x86_64-linux.so
> /usr/lib64/kodi/addons/library.kodi.peripheral/libKODI_peripheral-x86_64-linux.so
> /usr/lib64/kodi/addons/library.kodi.inputstream/libKODI_inputstream-x86_64-linux.so
> /usr/lib64/kodi/addons/library.kodi.guilib/libKODI_guilib-x86_64-linux.so
> /usr/lib64/kodi/addons/library.kodi.audioengine/libKODI_audioengine-x86_64-linux.so
> /usr/lib64/kodi/addons/library.kodi.adsp/libKODI_adsp-x86_64-linux.so
> /usr/lib64/kodi/system/libsse4-x86_64-linux.so
> /usr/lib64/kodi/kodi.bin
> 
> fix here:
> cjmayo/xbmc@244a462

Can you please submit that PR to kodi upstream? It seems like they'd want it and we don't want to carry that patch in Gentoo forever.

The rest I've done - thanks again for all the help

cjmayo commented:
I've submitted a PR https://github.com/xbmc/xbmc/pull/11213 to add a single option to try and replace both patches! We'll see how that goes.

Thanks for taking on getting Kodi updated in the tree. Kodi 16 didn't work for me and it was simpler to go straight to 17 rather than figure out exactly why - but that meant looking after my own ebuild which I'll be glad to get rid of.

hauihau commented:
@candrews 

some more cosmetics, you forget some sortings:
- in `COMMON_DEPEND`:
 -- `dvd?` should be after `dev-python/pillow` which should be after `dev-libs/yajl-2`
 -- `webserver?` should be after `vdpau`
- in `IUSE`:
 -- swap `gles` and `dvd`
 -- swap `nonfree` and `nfs` 

There have been lots of already merged commits which raise the version of dependencies in kodi master in the last time:
- xbmc/PR#11101 (not yet in portage)
- xbmc/PR#11102 (already in portage)
- xbmc/PR#11103 (already in portage)
- xbmc/PR#11104 (already in portage)
- xbmc/PR#11105 (already in portage)

Also some outstandig PRs:
- xbmc/PR#11100 (already in portage)
- xbmc/PR#11106 (not yet in portage)

Perhaps this should be included in the ebuild where appropriate.

And I still vote for an einfo when lirc is enabled. Are there any other opinions here on this?

candrews commented:
> xbmc/PR#11101 (not yet in portage)

reported https://bugs.gentoo.org/show_bug.cgi?id=603040

>     xbmc/PR#11102 (already in portage)

done

>     xbmc/PR#11103 (already in portage)

We're actually missing a dependency on media-libs/libjpeg-turbo - so I added that.

>     xbmc/PR#11104 (already in portage)

done

>     xbmc/PR#11105 (already in portage)

done

> Also some outstandig PRs:
> 
>     xbmc/PR#11100 (already in portage)

Since the current version requirement is net-libs/libmicrohttpd-0.9.50, that's what I put in the ebuild. When that changes, we can change the ebuild.

>     xbmc/PR#11106 (not yet in portage)

Since the current version requirement is dev-libs/openssl-1.0.2j, that's what I put in the ebuild. When that changes, we can change the ebuild. Also reported https://bugs.gentoo.org/show_bug.cgi?id=603044

> And I still vote for an einfo when lirc is enabled. Are there any other opinions here on this?

I'd love to hear another opinion on this too. @SoapGentoo, what do you think? We have a this runtime dependency:
```
        lirc? (
                || ( app-misc/lirc app-misc/inputlircd )
        )
```
Should we instead eliminate the dependency and add an `einfo` telling users if they want to use lirc to install it?
I'd also love your feedback on this ebuild in general, as I feel like it's mergeable (except for the media-libs/taglib-1.11.1 version bump not being in portage yet) at this point.

SoapGentoo commented on media-tv/kodi/kodi-9999.ebuild:
```diff
@@ -152,131 +148,78 @@ In some cases Kodi needs to access multicast addresses.
-	export ac_cv_lib_bluetooth_hci_devid=$(usex bluetooth)
-	# Requiring java is asine #434662
-	[[ ${PV} != 9999 ]] && export ac_cv_path_JAVA_EXE=$(which $(usex java java true))
+	local CMAKE_BUILD_TYPE=$(usex debug Debug Release)
```
Try and avoid this like the plague. `CMakeLists.txt` should be written such that it doesn't care about our `CMAKE_BUILD_TYPE=Gentoo`.

candrews replied:
Kodi uses the build type to decide if stripping should be done or not, as well as what optimizations to perform (see https://github.com/xbmc/xbmc/pull/11213#issuecomment-267912415 ). So I think we have to set CMAKE_BUILD_TYPE to either `Debug` or `RelWithDebInfo` and not `Gentoo` right?

SoapGentoo commented on media-tv/kodi/kodi-9999.ebuild:
```diff
@@ -152,131 +148,78 @@ In some cases Kodi needs to access multicast addresses.
+	local CMAKE_BUILD_TYPE=$(usex debug Debug Release)
+
+	local mycmakeargs=(
+		-Ddocdir=/usr/share/doc/${PF}
```
`EPREFIX`

candrews replied:
Done - thanks!

SoapGentoo commented on media-tv/kodi/kodi-9999.ebuild:
```diff
@@ -152,131 +148,78 @@ In some cases Kodi needs to access multicast addresses.
-		$(use_enable X x11) \
-		$(use_enable zeroconf avahi)
-}
+	use libusb && mycmakeargs+=( -DENABLE_LIBUSB=$(usex libusb) )
```
why is this all by itself here?

candrews replied:
If `-DENABLE_LIBUSB=` is provided when `-DENABLE_UDEV=OFF` a warning is emitted:
```
Manually-specified variables were not used by the project: ENABLE_LIBUSB
```
See https://github.com/gentoo/gentoo/pull/3027#issuecomment-266327985

SoapGentoo replied:
fine, thanks for stating it again :smile:

hauihau commented:
`dev-db/sqlite` should be before `dev-libs/expat`

>>xbmc/PR#11103 (already in portage)

> We're actually missing a dependency on media-libs/libjpeg-turbo - so I added that.

please move `media-libs/libjpeg-turbo` to DEPENDS and remove `virtual/jpeg:0=`

cjmayo commented:
A few more:

* Looks like x11-libs/libX11 is a direct dependency with USE="X"?:
```
$ grep -rn X11/Xlib.h
project/cmake/modules/FindX.cmake:22:find_path(X_INCLUDE_DIR NAMES X11/Xlib.h
xbmc/cores/VideoPlayer/DVDCodecs/Video/VDPAU.h:48:#include <X11/Xlib.h>
xbmc/powermanagement/DPMSSupport.cpp:121:#include <X11/Xlib.h>
xbmc/video/videosync/VideoSyncGLX.h:27:#include <X11/Xlib.h>
xbmc/windowing/X11/GLContext.h:24:#include "X11/Xlib.h"
xbmc/windowing/X11/WinSystemX11.h:31:#include "X11/Xlib.h"
xbmc/windowing/X11/WinSystemX11GLContext.cpp:24:#include <X11/Xlib.h>
xbmc/windowing/WinEventsX11.cpp:30:#include <X11/Xlib.h>
xbmc/windowing/WinEventsX11.h:24:#include <X11/Xlib.h>
xbmc/windowing/WinEventsSDL.cpp:40:#include <X11/Xlib.h>
xbmc-xrandr.c:29:#include <X11/Xlib.h>
```

* Possibly media-libs/mesa[egl] for some configurations at least?:
`-- Found EGL: /usr/lib64/libEGL.so (found version "13.0.2")`
and docs/README.linux: `Build-Depends: libegl1-mesa-dev`
```
$ grep -rn EGL/egl.h
xbmc/cores/VideoPlayer/DVDCodecs/Video/OpenMaxVideo.h:27:#include <EGL/egl.h>
xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.h:23:#include <EGL/egl.h>
xbmc/cores/omxplayer/OMXImage.h:38:#include <EGL/egl.h>
xbmc/windowing/X11/GLContextEGL.h:25:#include "EGL/egl.h"
xbmc/windowing/X11/WinSystemX11GLContext.h:27:#include "EGL/egl.h"
xbmc/windowing/X11/WinSystemX11GLESContext.h:28:#include "EGL/egl.h"
xbmc/windowing/egl/EGLNativeTypeIMX.cpp:24:#include <EGL/egl.h>
xbmc/windowing/egl/EGLNativeTypeIMX.h:28:#include <EGL/egl.h>
xbmc/windowing/egl/EGLWrapper.h:27:#include <EGL/egl.h>
xbmc/windowing/egl/WinSystemEGL.h:31:#include <EGL/egl.h>
xbmc/windowing/egl/EGLNativeTypeAndroid.cpp:23:#include <EGL/egl.h>
xbmc/windowing/egl/EGLNativeTypeRaspberryPI.cpp:22:#include <EGL/egl.h>
xbmc/windowing/egl/EGLNativeTypeAmlogic.cpp:32:#include <EGL/egl.h>
xbmc/windowing/mir/GLContextEGL.h:25:#include "EGL/egl.h"
```

My take on `/usr/bin/TexturePacker`:
* It's ~ 80KB. For comparison media-fonts/noto on its own is circa 16MB, and we only need a fraction of that
* Personally I'm building on the machine I'm running on so libs png, gif and jpeg are installed. They are likely needed by other things anyway
* Gentoo should make it easy for people to be able to develop their own skins
* Extra flags, separate ebuilds are just too much hassle ... for 80KB!

candrews commented:
@cjmayo I've made those changes - you're finding lots of really good stuff!

Regarding TexturePacker, I agree - for 80kb, it's just not worth the effort. So I suggest we do as you suggest - which is do nothing :-)

candrews commented:
I'm trying to decide what to do about binary addons... I'm having a tough time even figuring out how to build them: `cmake-utils_src_compile binary-addons` in `src_compile` does it, but Kodi downloads a bunch of stuff which is prohibited by Gentoo QA (`FEATURES="network-sandbox"`), so I need to eliminate all that downloading.

It also appears each binary addons can have different dependencies than kodi itself.

I'm wondering if use flags for each binary addons are the way to go, or if each should be it's own ebuild.

Anyone have any thoughts?

candrews commented:
I just added kodi-17.0_beta7.ebuild which is the same as kodi-9999.ebuild but adds one patch for https://github.com/xbmc/xbmc/pull/11088 - @wsnipex is it possible for you to backport that fix to Krypton so it's in the next 17.0-whatever release?

hauihau commented:
Regarding EGL:

According to `project/cmake/CMakeLists.txt` EGL is a requirement when `NOT WIN32` and `NOT APPLE`. So there should be a explicit dep on `media-libs/mesa[egl]` in `DEPEND`  (the dep behind `gles?` is not sufficient and can be removed). 

Also `media-libs/opengl-apple` does not look like its providing EGL, so the  `opengl?` block in `DEPEND` should be changed from

```
    opengl? (
        virtual/glu
        virtual/opengl
    )
```
to

```
    opengl? (
        media-libs/glu
        media-libs/mesa
    )
```

frace commented:
> I'm wondering if use flags for each binary addons are the way to go, or if each should be it's own ebuild.
> Anyone have any thoughts?

Hi all,

I created a bug report a while back and we had some initial discussions about that topic, see [555792](https://bugs.gentoo.org/show_bug.cgi?id=555792).
The discussion over there became stale for some time now. Personally I hadn't much time left during the last months to care about it. Same applies to my [overlay](https://github.com/frace/kodi-overlay). Btw. I'll try to get `net-libs/shairplay` into the tree again by starting a pull request here on Github...

There's also an overlay containing ebuilds and an eclass for - as far as I can tell - every binary addon, which I did not test: [gentoo-kodi-overlay](https://github.com/herrnst/gentoo-kodi-overlay)

Personally I'd follow the atomic strategy of creating a single ebuild for each addon instead of integrating them into the kodi ebuild. Other projects take this route, too. There's also a category for this type of ebuilds. Just have a look at the `media-plugins/*` folder. Maybe - as an alternative approach - it would work well to group addons by category and create addon ebuilds for these categories. But I assume that could complicate things and possibly wouldn't be very *KISS*. :)

wsnipex commented:
The whole reason for binary addons is to decouple them for kodi core, so separate ebuilds for each addon is recommended.
Note that @herrnst, which provides the above overlay is a team kodi member.

herrnst commented:
Chiming in as I got pinged here - If there's interest and/or I can help bring the addon ebuilds into portage (maybe even versioned ones for Krypton once it's marked stable/final), let me know. But note I didn't submit anything to portage yet. Of course you may also pick anything from my overlay for submission.

Note: The eclass is simply there to not have to repeat -DCMAKE_INSTALL_LIBDIR=$(get_libdir)/kodi a few dozen times for every addon that lives in media-plugins/kodi-*, which is required to have everything installed at the right location.

candrews commented:
I absolutely think we should set a goal of getting the binary add-ons in portage. Having both live and verisoned ebuilds would be great - that would make life far easier for both users and maintainers.

@herrnst for next steps, I'm thinking I'll copy the binary add-ons ebuilds from your overlay at https://github.com/herrnst/gentoo-kodi-overlay into this PR. Then hopefully we can get this merged.

When Kodi 17 final is released, you or I can or someone else can submit PRs to version the add-on ebuilds.

I think the goal should be to eliminate your overlay. (No offense, but having all this work upstreamed is best for everyone 👍 )

herrnst commented:
Feel free to pick and upstream whatever is wanted/needed ;) Not entirely sure if the eclass is really needed, but it does ease and unclutter things a bit. Regarding versioning: It might be needed to reference addons by their GIT commits.

candrews commented:
@herrnst after Kodi 17 final is released, the addons will release numbered versions and corresponding tarballs I assume?

herrnst commented:
@candrews They're packaged and distributed according to what is referenced in the Krypton branch of https://github.com/xbmc/repo-binary-addons.git which in case of git can basically be any reference (branch, tag, commit hash etc.), and packagers should build based on these references. There are no tarballs created. So when creating ebuilds, the correct sources/versions are probably best determined by their GIT commit hash (e.g. like it's currently done with the references to the external DVD libraries in the kodi ebuild). Normally, after a Kodi version is released, the addons for that specific version don't change very much (if at all), so there won't be much maintenance.

That said, I'm not very much into our full release process - as @wsnipex is our debian packager, please correct me if I'm wrong here.

herrnst commented:
Some additionals and a thought: Looking for example at https://github.com/xbmc/repo-binary-addons/blob/Jarvis/pvr.hts/pvr.hts.txt you see the reference to which version/HEAD is built and distributed for the pvr.hts addon in the Kodi Jarvis release (https://github.com/kodi-pvr/pvr.hts/tree/Jarvis). In the addon's GIT repo/branch, in pvr.hts/ there's addon.xml.in which carries the addon version in <addon version="xxx"> (2.2.20 at the moment). So a versioned ebuild might look like this (commit hash resolved from the Jarvis HEAD):

    # media-plugins/kodi-pvr-hts-2.2.20.ebuild
    ADDON_COMMIT="92610477abbf4784399964a180a1784aee319039"
    SRC_URI="https://github.com/kodi-pvr/pvr.hts/archive/${ADDON_COMMIT}.tar.gz -> kodi-pvr-hts-${PV}.tar.gz"

where PV might also be ADDON_COMMIT, but you get the idea. The resulting tarballs from SRC_URI can IMHO then live in /usr/portage/distfiles and on distfiles.gentoo.org etc.

candrews commented:
@herrnst that sounds like a great approach.

As I'm starting to work on this, the first question I have is why your overlay has dev-libs/libp8-platform. You split off from dev-libs/libplatform in https://github.com/herrnst/gentoo-kodi-overlay/commit/e7bb4993033d1b8905178b091a84c6ae1a2e6ec5 and I'm not sure why. I think if we can keep using dev-libs/libplatform (but bumping the version, of course) that would be ideal - but I clearly don't know enough. Can you help me understand?

candrews commented:
I'm tackling the addons at https://github.com/gentoo/gentoo/pull/3195.

So I think this PR is ready for merging. I've removed the [wip] indicator. Can we get this in?

herrnst commented:
@candrews re dev-libs/libplatform: The rename was done due to the upstream rename (which was neccessary due to some package naming conflict with debian IIRC), which also involved API breakage (namespace was moved from PLATFORM:: to P8PLATFORM::). Also, IIRC, there wasn't a live ebuild available for the platform lib at that time. Lastly, I wanted all Kodi related things in one place/overlay.

Since the platform lib really is p8platform now, IMHO the ebuild should be named accordingly and also conflict with the old library, and also have the package name aligned with other distros like debian/ubuntu.

herrnst commented:
@candrews Before this PR gets merged, you should pick https://github.com/herrnst/gentoo-kodi-overlay/commit/14917a6885a36cb07436aaa276c91a669fcd1b5e (or adjust accordingly) as the CMake project files were moved.

candrews commented:
Thanks @herrnst! I've made that change.

SoapGentoo commented on media-tv/kodi/kodi-17.0_rc1.ebuild:
```diff
@@ -0,0 +1,249 @@
+	>=net-misc/curl-7.51.0
+	nfs? ( net-fs/libnfs:= )
+	opengl? ( media-libs/glu )
+	openssl? ( >=dev-libs/openssl-1.0.2j:0= )
```
is this the only SSL/TLS/whatever provider? If so, `USE=ssl` enables/disables SSL/TLS, whereas `USE=openssl` selects the impl, conditional on `USE=ssl`

candrews replied:
libressl also works, so I've updated it accordingly.

SoapGentoo commented on media-tv/kodi/kodi-17.0_rc1.ebuild:
```diff
@@ -0,0 +1,249 @@
+}
+
+src_unpack() {
+	[[ ${PV} == 9999 ]] && git-r3_src_unpack || default
```
is this really necessary? doesn't `git-r3.eclass` override it if its inherited in a -9999 ebuild? Just asking

candrews replied:
Indeed it does! That's really cool

SoapGentoo commented on media-tv/kodi/kodi-17.0_rc1.ebuild:
```diff
@@ -0,0 +1,249 @@
+}
+
+src_prepare() {
+	default
```
never use `default` with cmake-utils, always use `cmake-utils_src_prepare`!

candrews replied:
Today I learned :-)

SoapGentoo commented on media-tv/kodi/kodi-17.0_rc1.ebuild:
```diff
@@ -0,0 +1,249 @@
+}
+
+src_configure() {
+	local CMAKE_BUILD_TYPE=$(usex debug Debug RelWithDebInfo)
```
in general please totally avoid this. Gentoo sets its own `CMAKE_BUILD_TYPE` which is totally orthogonal to the classic ones, and if the build breaks, it makes too many assumptions about `CMAKE_BUILD_TYPE`. CMake-based build systems should be agnostic of `CMAKE_BUILD_TYPE`.

candrews replied:
While I agree in principal, Kodi only works with specific cmake build types; it fails if the cmake build type is set to `Gentoo` as happens by default. So I think we have to set `CMAKE_BUILD_TYPE` here, unfortunately.

SoapGentoo replied:
alright then, can you try and get this removed upstream then?

SoapGentoo commented on media-tv/kodi/kodi-17.0_rc1.ebuild:
```diff
@@ -0,0 +1,249 @@
+}
+
+src_compile() {
+	default
```
call `cmake-utils_src_compile` instead of `default`

candrews replied:
👍

SoapGentoo commented on media-tv/kodi/kodi-17.0_rc1.ebuild:
```diff
@@ -0,0 +1,249 @@
+
+	rm "${ED%/}"/usr/share/kodi/addons/skin.estuary/fonts/Roboto-Thin.ttf || die
+	dosym /usr/share/fonts/roboto/Roboto-Thin.ttf \
+		usr/share/kodi/addons/skin.estuary/fonts/Roboto-Thin.ttf
```
I haven't compared them bitwise, but can't this be done in a loop to reduce the churn? Just a suggestion

candrews commented on media-tv/kodi/kodi-17.0_rc1.ebuild:
```diff
@@ -0,0 +1,249 @@
+	|| ( gles opengl )
+	udev? ( !libusb )
+	udisks? ( dbus )
+	upower? ( dbus )
```
That's obvious... I wonder if repoman can be improved to catch that kind of thing.

frace commented:
Isn't it recommended to use eclass functions provided by inheritance?

E.g. from [CMAKE-UTILS.ECLASS](https://devmanual.gentoo.org/eclass-reference/cmake-utils.eclass/):

> cmake-utils_use_enable <USE flag> [flag name]
> Based on use_enable. See ebuild(5).
> `cmake-utils_use_enable foo FOO` echoes -DENABLE_FOO=ON if foo is enabled and -DENABLE_FOO=OFF if it is disabled.

candrews commented:
> Isn't it recommended to use eclass functions provided by inheritance?
> 
> E.g. from CMAKE-UTILS.ECLASS:
> 
>     cmake-utils_use_enable [flag name]
>     Based on use_enable. See ebuild(5).
>     cmake-utils_use_enable foo FOO echoes -DENABLE_FOO=ON if foo is enabled and -DENABLE_FOO=OFF if it is disabled.

cmake-utils_use_* is banned in EAPI 6: https://groups.google.com/forum/#!topic/linux.gentoo.dev/QxCQJ_8taqg

frace commented:
> cmake-utils_use_* is banned in EAPI 6: https://groups.google.com/forum/#!topic/linux.gentoo.dev/QxCQJ_8taqg

I see the point of the discussion. But I can't see the final decision to do so. According to the current version and documentation it still exists:
+ [Source](https://github.com/gentoo/gentoo/blob/master/eclass/cmake-utils.eclass)
+ [Documentation](https://devmanual.gentoo.org/eclass-reference/cmake-utils.eclass/)

SoapGentoo commented:
yes, all those eclass functions have been killed because they add no value

SoapGentoo commented on media-tv/kodi/kodi-9999.ebuild:
```diff
@@ -289,19 +225,13 @@ src_install() {
-	rm "${ED%/}"/usr/share/kodi/addons/skin.estuary/fonts/NotoSans-Regular.ttf || die
-	dosym /usr/share/fonts/noto/NotoSans-Regular.ttf \
-		usr/share/kodi/addons/skin.estuary/fonts/NotoSans-Regular.ttf
+	for f in NotoMono-Regular.ttf NotoSans-Bold.ttf NotoSans-Regular.ttf ; do
```
don't forget to make f `local`

SoapGentoo commented on media-tv/kodi/kodi-17.0_rc1.ebuild:
```diff
@@ -0,0 +1,242 @@
+	dosym /usr/share/fonts/noto/NotoSans-Regular.ttf \
+		usr/share/kodi/addons/skin.estouchy/fonts/NotoSans-Regular.ttf
+
+	for f in NotoMono-Regular.ttf NotoSans-Bold.ttf NotoSans-Regular.ttf ; do
```
make f `local`

SoapGentoo commented on media-tv/kodi/kodi-17.0_rc1.ebuild:
```diff
@@ -0,0 +1,242 @@
+
+	rm "${ED%/}"/usr/share/kodi/addons/skin.estouchy/fonts/Roboto-Thin.ttf || die
+	dosym /usr/share/fonts/roboto/Roboto-Thin.ttf \
+		usr/share/kodi/addons/skin.estuary/fonts/Roboto-Thin.ttf
```
can't you put everything in the loop with
```
	local f
	for f in NotoMono-Regular.ttf NotoSans-Bold.ttf NotoSans-Regular.ttf NotoSans-Regular.ttf Roboto-Thin.ttf; do
		rm "${ED%/}"/usr/share/kodi/addons/skin.*/fonts/"${f}" || die
		dosym /usr/share/fonts/noto/"${f}" \
			usr/share/kodi/addons/skin.estuary/fonts/"${f}"
	done
```

candrews replied:
I don't think so.
* "Roboto-Thin.ttf" isn't in /usr/share/fonts/noto/ so `dosym` would make a broken symlink
* There are two skins, esuary and estouchy - this loop would remove the fonts from both estuary and estouchy, but only add them back into estuary.

frace commented:
> Agreed - here's the bug for shairplay in Gentoo: https://bugs.gentoo.org/show_bug.cgi?id=538932
If it ever gets into portage, we can revisit.

Just wanted to let you know I opened a PR to get `net-libs/shairplay` into portage: #3425
